js/src/components/paid-ads/asset-group/asset-group.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { useAdaptiveFormContext } from '.~/components/adaptive-form';
+import StepContent from '.~/components/stepper/step-content';
+import StepContentHeader from '.~/components/stepper/step-content-header';
+import StepContentFooter from '.~/components/stepper/step-content-footer';
+import AppButton from '.~/components/app-button';
+import AssetGroupSection from './asset-group-section';
+
+export const ACTION_COMPLETE = 'submit-complete';
+export const ACTION_SKIP_ASSETS = 'submit-with-skipping-assets';
+
+/**
+ * Renders the form container for managing an asset group of a campaign.
+ *
+ * Please note that this component relies on an AdaptiveForm's context, so it expects
+ * a context provider component (`AdaptiveForm`) to existing in its parents.
+ */
+export default function AssetGroup() {
+	const { isValidForm, handleSubmit, adapter } = useAdaptiveFormContext();
+	const { isSubmitting, isSubmitted, submitter } = adapter;
+	const currentAction = submitter?.dataset.action;
+
+	return (
+		<StepContent>
+			<StepContentHeader
+				title={ __( 'Boost your campaign', 'google-listings-and-ads' ) }
+				description={ __(
+					'Get more conversions by adding creative assets to your campaign',
+					'google-listings-and-ads'
+				) }
+			/>
+			<AssetGroupSection />
+			<StepContentFooter>
+				<AppButton
+					isTertiary
+					data-action={ ACTION_SKIP_ASSETS }
+					disabled={
+						// TODO: Change to only validate campaign data.
+						! isValidForm ||
+						isSubmitted ||
+						currentAction === ACTION_COMPLETE
+					}
+					loading={
+						isSubmitting && currentAction === ACTION_SKIP_ASSETS
+					}
+					onClick={ handleSubmit }
+				>
+					{ __( 'Skip adding assets', 'google-listings-and-ads' ) }
+				</AppButton>
+				<AppButton
+					isPrimary
+					data-action={ ACTION_COMPLETE }
+					disabled={
+						! isValidForm ||
+						isSubmitted ||
+						currentAction === ACTION_SKIP_ASSETS
+					}
+					loading={
+						isSubmitting && currentAction === ACTION_COMPLETE
+					}
+					onClick={ handleSubmit }
+				>
+					{ __( 'Launch paid campaign', 'google-listings-and-ads' ) }
```

Comment:
💅 
Maybe this is because of `// TODO: Change to only validate campaign data.` But if not... with a little tweak in the disabled logic we can have a potential refactor here:

i.e.
```
<StepFooterButton action={ACTION_SKIP_ASSETS} isTerciary>{ __( 'Skip adding assets', 'google-listings-and-ads' ) }</StepFooterButton>
<StepFooterButton action={ACTION_COMPLETE} isPrimary>{ __( 'Launch paid campaign', 'google-listings-and-ads' ) }</StepFooterButton>
```

```
[...]
data-action={ action }
disabled={
						! isValidForm ||
						isSubmitted ||
						currentAction !== action
					}
[...]
```

Reply:
Do you mean a code like this?

```js
export default function AssetGroup() {
	const { isValidForm, handleSubmit, adapter } = useAdaptiveFormContext();
	const { isSubmitting, isSubmitted, submitter } = adapter;
	const currentAction = submitter?.dataset.action;

	function StepFooterButton( { action, ...props } ) {
		return (
			<AppButton
				{ ...props }
				loading={ isSubmitting && currentAction === action }
				data-action={ action }
				onClick={ handleSubmit }
			/>
		);
	}

	return (
		<StepContent>
			[ ... omitted ]
			<StepContentFooter>
				<StepFooterButton
					isTertiary
					action={ ACTION_SKIP_ASSETS }
					disabled={
						// TODO: Change to only validate campaign data.
						! isValidForm ||
						isSubmitted ||
						currentAction === ACTION_COMPLETE
					}
				>
					{ __( 'Skip adding assets', 'google-listings-and-ads' ) }
				</StepFooterButton>
				<StepFooterButton
					isPrimary
					action={ ACTION_COMPLETE }
					disabled={
						! isValidForm ||
						isSubmitted ||
						currentAction === ACTION_SKIP_ASSETS
					}
				>
					{ __( 'Launch paid campaign', 'google-listings-and-ads' ) }
				</StepFooterButton>
			</StepContentFooter>
		</StepContent>
	);
}
```

If so, however, I'm not pretty sure it's worth it. Because button implementations are instead split into shattered code fragments, which reduces readability and doesn't improve maintainability much.

Reply:
Something like this. Thats why I asked about that TODO thing

```jsx
export default function AssetGroup() {
	const { isValidForm, handleSubmit, adapter } = useAdaptiveFormContext();
	const { isSubmitting, isSubmitted, submitter } = adapter;
	const currentAction = submitter?.dataset.action;

	function StepFooterButton( { action, ...props } ) {
		return (
			<AppButton
				{ ...props }
				loading={ isSubmitting && currentAction === action }
				data-action={ action }
				onClick={ handleSubmit }
                                disabled={
						! isValidForm ||
						isSubmitted ||
						currentAction !== action
					}
			/>
		);
	}

	return (
		<StepContent>
			[ ... omitted ]
			<StepContentFooter>
				<StepFooterButton
					isTertiary
					action={ ACTION_SKIP_ASSETS }
				>
					{ __( 'Skip adding assets', 'google-listings-and-ads' ) }
				</StepFooterButton>
				<StepFooterButton
					isPrimary
					action={ ACTION_COMPLETE }
				>
					{ __( 'Launch paid campaign', 'google-listings-and-ads' ) }
				</StepFooterButton>
			</StepContentFooter>
		</StepContent>
	);
}
```

Reply:
Thanks for the explanation.

> Maybe this is because of `// TODO: Change to only validate campaign data.` [...]

> Thats why I asked about that TODO thing

Sorry, I didn't fully understand what you meant originally.

Yes, since that submit button is used to create a new campaign without assets, it will be changed to validate `countryCodes` and `amount` only in a later PR, along with changing the validation in *utils/paid-ads/validateForm.js* to also check asset data.

> But if not... with a little tweak in the disabled logic we can have a potential refactor here: [...]

`currentAction === ACTION_COMPLETE` and `currentAction === ACTION_SKIP_ASSETS` are not exactly logically equivalent to `currentAction !== action` since the `currentAction` could be `undefined`, but using `Boolean( currentAction && currentAction !== action )` could solve this.

---

Taken together, the code could be like this:

```js
export default function AssetGroup() {
	const { isValidForm, handleSubmit, adapter } = useAdaptiveFormContext();
	const { isSubmitting, isSubmitted, submitter } = adapter;
	const currentAction = submitter?.dataset.action;

	function StepFooterButton( { action, disabledByValidation, ...props } ) {
		return (
			<AppButton
				{ ...props }
				loading={ isSubmitting && currentAction === action }
				disabled={
					disabledByValidation ||
					isSubmitted ||
					Boolean( currentAction && currentAction !== action )
				}
				data-action={ action }
				onClick={ handleSubmit }
			/>
		);
	}

	return (
		<StepContent>
			[ ... omitted ]
			<StepContentFooter>
				<StepFooterButton
					isTertiary
					action={ ACTION_SKIP_ASSETS }
					disabledByValidation={
						// TODO: Change to only validate campaign data.
						! isValidForm
					}
				>
					{ __( 'Skip adding assets', 'google-listings-and-ads' ) }
				</StepFooterButton>
				<StepFooterButton
					isPrimary
					action={ ACTION_COMPLETE }
					disabledByValidation={ ! isValidForm }
				>
					{ __( 'Launch paid campaign', 'google-listings-and-ads' ) }
				</StepFooterButton>
			</StepContentFooter>
		</StepContent>
	);
}
```

Visually, the shattered code fragments in implementation are a little better, but the number of props being split is still not better, and there is even another layer of logic to make the final `disabled`.

However, the current implementation is concentrated but looks trivial, so I don't think any of it is particularly good. If you prefer the splitting one, I'd be happy to adjust it.